0-module-and-module-source.emu

```diff
@@ -0,0 +1,661 @@
+          <td>
+            a List of ExportEntry Records
+          </td>
+          <td>
+            A List of ExportEntry records derived from the code of this module source that correspond to reexported imports that occur within the module or exports from `export * as namespace` declarations.
+          </td>
+        </tr>
+        <tr>
+          <td>
+            [[StarExportEntries]]
+          </td>
+          <td>
+            a List of ExportEntry Records
+          </td>
+          <td>
+            A List of ExportEntry records derived from the code of this module source that correspond to `export *` declarations that occur within the module, not including `export * as namespace` declarations.
+          </td>
+        </tr>
+        <tr>
+            <td>
+              [[HasTLA]]
+            </td>
+            <td>
+              a Boolean
+            </td>
+            <td>
+              Whether this module source is individually asynchronous. Having an asynchronous dependency does not mean this field is *true*.
+            </td>
+        </tr>
+        <tr>
+            <td>
+              [[RequestedModules]]
+            </td>
+            <td>
+              a List of Strings
+            </td>
+            <td>
+              A List of all the |ModuleSpecifier| strings used by the module source represented by this record to request the importation of a module. The List is source text occurrence ordered.
+            </td>
+        </tr>
+        <tr>
+            <td>
+              [[HostDefined]]
+            </td>
+            <td>
+              anything (default value is *undefined*)
+            </td>
+            <td>
+              Field reserved for use by host environments that need to associate additional information with a module source.
+            </td>
+          </tr>
+      </table>
+    </emu-table>
+
+    <emu-note type="editor">
+        A Module Source Record is the static portion of a Module Record. A refactor of the spec can make this separation more formal instead of just copying internal slot from one record to another.
+     </emu-note>
+</emu-clause>
+
+<emu-clause id="sec-import-calls">
+    <h1>Import Calls</h1>
+
+    <emu-clause id="sec-import-call-runtime-semantics-evaluation" type="sdo">
+      <h1>Runtime Semantics: Evaluation</h1>
+
+      <emu-grammar>ImportCall : `import` `(` AssignmentExpression `)`</emu-grammar>
+      <emu-alg>
+        1. Let _referencingScriptOrModule_ be GetActiveScriptOrModule().
+        1. Let _argRef_ be the result of evaluating |AssignmentExpression|.
+        1. <del>Let _specifier_ be ? GetValue(_argRef_).</del>
+        1. <ins>Let _specifierOrModule_ be ? GetValue(_argRef_).</ins>
+        1. Let _promiseCapability_ be ! NewPromiseCapability(%Promise%).
+        1. <ins>If Type(_specifierOrModule_) is Object that has a [[ModuleSourceInstance]] internal slot, then</ins>
+          1. <ins>Let _moduleRecord_ be _specifierOrModule_.[[Module]].</ins>
```

Comment:
Also, should we also be looking up if `_referencingScriptOrModule_` is itself a `SourceTextModule` (pretty sure we can assume it always be?) that we should look up its `[[Instance]]` slot in turn for the parent reference if there is one? Would that be useful information for an import hook even at the top level?

Reply:
> Should this be `[[Module]]` or `[[ModuleSourceInstance]]`?

It should be `[[Module]]` because the instance provided must already have a qualifying module record associate to it, and that's the currency of all these operations.

Reply:
> We may want to make that invariant explicit.

Any suggestion on how to make that more explicit?